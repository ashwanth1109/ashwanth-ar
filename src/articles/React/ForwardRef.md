## Forward Ref

Ref forwarding is a technique for automatically passing a ref through a component to one of its children. It is an opt-in feature that lets some components take a ref they receive, and pass it further down (in other words, “forward” it) to a child.

```js
const FancyButton = React.forwardRef((props, ref) => (
  <button ref={ref} className="FancyButton">
    {props.children}
  </button>
));

// You can now get a ref directly to the DOM button:
const ref = React.createRef();
<FancyButton ref={ref}>Click me!</FancyButton>;
```

The second ref argument only exists when you define a component with React.forwardRef call. Regular function or class components don’t receive the ref argument, and ref is not available in props either.

<hr/>

### Refs in HOC

Refs will not get passed through a HOC along with props. That’s because ref is not a prop. Like key, it’s handled differently by React. If you add a ref to a HOC, the ref will refer to the outermost container component, not the wrapped component.

```js
function logProps(Component) {
  class LogProps extends React.Component {
    componentDidUpdate(prevProps) {
      console.log("old props:", prevProps);
      console.log("new props:", this.props);
    }

    render() {
      const { forwardedRef, ...rest } = this.props;

      // Assign the custom prop "forwardedRef" as a ref
      return <Component ref={forwardedRef} {...rest} />;
    }
  }

  // Note the second param "ref" provided by React.forwardRef.
  // We can pass it along to LogProps as a regular prop, e.g. "forwardedRef"
  // And it can then be attached to the Component.
  return React.forwardRef((props, ref) => {
    return <LogProps {...props} forwardedRef={ref} />;
  });
}
```

We can explicitly forward refs to the inner FancyButton component using the React.forwardRef API. React.forwardRef accepts a render function that receives props and ref parameters and returns a React node.

<hr/>

### SOURCE CODE - forwardRef.js

```js
import { REACT_FORWARD_REF_TYPE, REACT_MEMO_TYPE } from "shared/ReactSymbols";

import warningWithoutStack from "shared/warningWithoutStack";

export default function forwardRef<Props, ElementType: React$ElementType>(
  render: (props: Props, ref: React$Ref<ElementType>) => React$Node
) {
  if (__DEV__) {
    if (render != null && render.$$typeof === REACT_MEMO_TYPE) {
      warningWithoutStack(
        false,
        "forwardRef requires a render function but received a `memo` " +
          "component. Instead of forwardRef(memo(...)), use " +
          "memo(forwardRef(...))."
      );
    } else if (typeof render !== "function") {
      warningWithoutStack(
        false,
        "forwardRef requires a render function but was given %s.",
        render === null ? "null" : typeof render
      );
    } else {
      warningWithoutStack(
        // Do not warn for 0 arguments because it could be due to usage of the 'arguments' object
        render.length === 0 || render.length === 2,
        "forwardRef render functions accept exactly two parameters: props and ref. %s",
        render.length === 1
          ? "Did you forget to use the ref parameter?"
          : "Any additional parameter will be undefined."
      );
    }

    if (render != null) {
      warningWithoutStack(
        render.defaultProps == null && render.propTypes == null,
        "forwardRef render functions do not support propTypes or defaultProps. " +
          "Did you accidentally pass a React component?"
      );
    }
  }

  return {
    $$typeof: REACT_FORWARD_REF_TYPE,
    render
  };
}
```

<hr/>
<br/>
<br/>

```js
if (render != null && render.$$typeof === REACT_MEMO_TYPE) {
  warningWithoutStack(
    false,
    "forwardRef requires a render function but received a `memo` " +
      "component. Instead of forwardRef(memo(...)), use " +
      "memo(forwardRef(...))."
  );
}
```

Check if there is something to render and if its a memoized component, then throw out a warning message. So, basically pass in a forwardedRef component to be memoized and not the other way around.

<hr/>

```js
if (typeof render !== "function") {
  warningWithoutStack(
    false,
    "forwardRef requires a render function but was given %s.",
    render === null ? "null" : typeof render
  );
}
```

Check for render function.

<hr/>

```js
warningWithoutStack(
  // Do not warn for 0 arguments because it could be due to usage of the 'arguments' object
  render.length === 0 || render.length === 2,
  "forwardRef render functions accept exactly two parameters: props and ref. %s",
  render.length === 1
    ? "Did you forget to use the ref parameter?"
    : "Any additional parameter will be undefined."
);
```

Check if render function has 0 or 2 parameters. If not, then throw appropriate warnings.

<hr/>

```js
warningWithoutStack(
  render.defaultProps == null && render.propTypes == null,
  "forwardRef render functions do not support propTypes or defaultProps. " +
    "Did you accidentally pass a React component?"
);
```

Does this mean that when you use forwardRef, you cannot use propTypes or defaultProps? That doesnt make sense?!!
