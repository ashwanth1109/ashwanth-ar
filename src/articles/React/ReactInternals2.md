# A Deep Dive into React Internals - Part 2

## createRef

```jsx
// an immutable object with a single mutable value
export function createRef(): RefObject {
  const refObject = {
    current: null
  };
  if (__DEV__) {
    Object.seal(refObject);
  }
  return refObject;
}
```

`__DEV__` => global variable to track whether development or production

`Object.seal()` => This method seals an object preventing new properties from being added to it and marking all existing properties as non-configurable. Values of present properties can still be changed as long as they are writable.

```jsx
export type RefObject = {|
  current: any
|};
```

## ReactCreateEventComponent

```js
import type {
  ReactEventComponent,
  ReactEventResponder
} from "shared/ReactTypes";
import { enableEventAPI } from "shared/ReactFeatureFlags";

import { REACT_EVENT_COMPONENT_TYPE } from "shared/ReactSymbols";
```

### What are polyfills? (from MDN docs)

A polyfill is a piece of code (usually JavaScript on the Web) used to provide modern functionality on older browsers that do not natively support it.

The reason for why polyfills are not used exclusively is for better functionality and better performance. Native implementations of APIs can do more and are faster than polyfills.

```js
let hasBadMapPolyfill;
```

```js
if (__DEV__) {
  hasBadMapPolyfill = false;
  try {
    const frozenObject = Object.freeze({});
    const testMap = new Map([[frozenObject, null]]);
    const testSet = new Set([frozenObject]);
    // This is necessary for Rollup to not consider these unused.
    // https://github.com/rollup/rollup/issues/1771
    // TODO: we can remove these if Rollup fixes the bug.
    testMap.set(0, 0);
    testSet.add(0);
  } catch (e) {
    // TODO: Consider warning about bad polyfills
    hasBadMapPolyfill = true;
  }
}
```

Rollup is a module bundler for JavaScript which compiles small pieces of code into something larger and more complex, such as a library or application.

```js
export function createEventComponent(
  responder: ReactEventResponder,
  displayName: string
): ?ReactEventComponent {
  if (enableEventAPI) {
    // We use responder as a Map key later on. When we have a bad
    // polyfill, then we can't use it as a key as the polyfill tries
    // to add a property to the object.
    if (__DEV__ && !hasBadMapPolyfill) {
      Object.freeze(responder);
    }
    const eventComponent = {
      $$typeof: REACT_EVENT_COMPONENT_TYPE,
      displayName: displayName,
      props: null,
      responder: responder
    };
    if (__DEV__) {
      Object.freeze(eventComponent);
    }
    return eventComponent;
  }
}
```

## ReactChildren

```js
import invariant from "shared/invariant";
import warning from "shared/warning";
import {
  getIteratorFn,
  REACT_ELEMENT_TYPE,
  REACT_PORTAL_TYPE
} from "shared/ReactSymbols";

import { isValidElement, cloneAndReplaceKey } from "./ReactElement";
import ReactDebugCurrentFrame from "./ReactDebugCurrentFrame";
```

What are these separators for?

```js
const SEPARATOR = ".";
const SUBSEPARATOR = ":";
```

Also, what key are they trying to escape here?

Probably similar to the key we provide when mapping through an array to generate html elements for each item in the array.

```js
/**
 * Escape and wrap key so it is safe to use as a reactid
 *
 * @param {string} key to be escaped.
 * @return {string} the escaped key.
 */
function escape(key) {
  const escapeRegex = /[=:]/g;
  const escaperLookup = {
    "=": "=0",
    ":": "=2"
  };
  const escapedString = ("" + key).replace(escapeRegex, function(match) {
    return escaperLookup[match];
  });

  return "$" + escapedString;
}
```

### To be continued
