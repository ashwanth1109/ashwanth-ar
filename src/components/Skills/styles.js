// @flow

import styled from "styled-components";

import { fonts } from "styles/constants";

export const SkillSection = styled.div`
  margin: 24px 0;
  width: 100%;
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
  grid-row-gap: 24px;
  grid-column-gap: 12px;
`;

export const IconContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const Icon = styled.img`
  height: 100px;
  max-width: 100px;
`;

export const IconTitle = styled.div`
  margin-top: 12px;
  display: flex;
  justify-content: center;
  text-transform: uppercase;
  font-family: ${fonts.primary};
  font-size: 24px;
`;

export const IconCell = styled.div``;
