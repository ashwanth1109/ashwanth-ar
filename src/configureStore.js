// @flow

import { applyMiddleware, combineReducers, compose, createStore } from "redux";
import { connectRoutes } from "redux-first-router";

import page from "./pageReducer";

const routesMap = {
  HOME: "/",
  ABOUT: "/about",
  TECH: "/tech",
  MUSIC: "/music",
  BOOKS: "/books"
};

export default function configureStore(preloadedState) {
  const { reducer, middleware, enhancer } = connectRoutes(routesMap);

  const rootReducer = combineReducers({ page, location: reducer });
  const middlewares = applyMiddleware(middleware);

  // For devtools extension
  const composeEnhancers =
    typeof window === "object" && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
      : compose;

  const enhancers = composeEnhancers(enhancer, middlewares);

  const store = createStore(rootReducer, preloadedState, enhancers);

  return { store };
}
