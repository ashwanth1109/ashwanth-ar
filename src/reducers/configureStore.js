// @flow

import { applyMiddleware, combineReducers, compose, createStore } from "redux";
import { connectRoutes } from "redux-first-router";

import page from "./pageReducer";

import app from "reducers";

const routesMap = {
  HOME: "/",
  ABOUT: "/about",
  TECH: "/tech",
  UNIFIZE: "/tech/unifize-chat",
  TECH_ARTICLES: "/tech/:article",
  MUSIC: "/music",
  MUSIC_ARTICLES: "/music/:article",
  CREDO: "/credo"
};

export default function configureStore(preloadedState) {
  const { reducer, middleware, enhancer } = connectRoutes(routesMap);

  const rootReducer = combineReducers({ page, location: reducer, app });
  const middlewares = applyMiddleware(middleware);

  // For devtools extension
  const composeEnhancers =
    typeof window === "object" && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
      : compose;

  const enhancers = composeEnhancers(enhancer, middlewares);

  const store = createStore(rootReducer, preloadedState, enhancers);

  return { store };
}
