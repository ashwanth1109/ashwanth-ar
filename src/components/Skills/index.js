// @flow

import React from "react";

import {
  SkillSection,
  Icon,
  IconContainer,
  IconTitle,
  IconCell
} from "./styles";
import { Container, Heading } from "styles";
import { iconPack1, iconPack2 } from "assets/icons";

const Skills = () => {
  return (
    <Container flex="column" margin="16px">
      <Heading>THINGS I KNOW PRETTY WELL</Heading>
      <SkillSection>
        {iconPack1.map((icon, id) => (
          <IconCell key={id}>
            <IconContainer>
              <Icon src={icon.src} />
            </IconContainer>
            <IconTitle>{icon.name}</IconTitle>
          </IconCell>
        ))}
      </SkillSection>
      <Heading>THINGS I THINK I KNOW</Heading>
      <SkillSection>
        {iconPack2.map((icon, id) => (
          <IconCell key={id}>
            <IconContainer>
              <Icon src={icon.src} />
            </IconContainer>
            <IconTitle>{icon.name}</IconTitle>
          </IconCell>
        ))}
      </SkillSection>
    </Container>
  );
};

export default Skills;
