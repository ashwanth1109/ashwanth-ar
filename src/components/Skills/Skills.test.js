// @flow

import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";

import Skills from "./";
import configureStore from "reducers/configureStore";

const { act } = renderer;
const { store } = configureStore();

test("Skills renders correctly", () => {
  let component;

  act(() => {
    component = renderer.create(
      <Provider store={store}>
        <Skills />
      </Provider>
    );
  });

  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
