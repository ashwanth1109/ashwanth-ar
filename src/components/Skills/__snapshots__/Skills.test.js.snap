// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skills renders correctly 1`] = `
<div
  className="sc-bZQynM fAcEir"
>
  <div
    className="sc-dnqmqq fThudL"
  >
    THINGS I KNOW PRETTY WELL
  </div>
  <div
    className="sc-bdVaJa jAKVfj"
  >
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="html.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        html5
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="css.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        css3
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="js.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        javascript
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="react.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        react
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="redux.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        redux
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="firebase.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        firebase
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="saga.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        saga
      </div>
    </div>
  </div>
  <div
    className="sc-dnqmqq fThudL"
  >
    THINGS I THINK I KNOW
  </div>
  <div
    className="sc-bdVaJa jAKVfj"
  >
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="git.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        git
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="node.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        node
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="express.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        express
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="mongo.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        mongo
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="next.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        next
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="jquery.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        jquery
      </div>
    </div>
    <div
      className="sc-ifAKCX cFlEyZ"
    >
      <div
        className="sc-bwzfXH cusHsC"
      >
        <img
          className="sc-htpNat hdMUAD"
          src="sketch.svg"
        />
      </div>
      <div
        className="sc-bxivhb gphwzh"
      >
        sketch
      </div>
    </div>
  </div>
</div>
`;
